libs/payments/ui/src/lib/nestapp/nextjs-actions.service.ts

```diff
@@ -343,4 +345,50 @@ export class NextJSActionsService {
       currency,
     };
   }
+
+  @SanitizeExceptions()
+  @NextIOValidator(DetermineLocationActionArgs, DetermineLocationActionResult)
+  @WithTypeCachableAsyncLocalStorage()
+  async determineLocation(args: {
+    sanctionedLocations: string;
+    offeringId: string;
+    acceptLanguage?: string | null;
+    selectedLanguage?: string;
+    countryCode?: string;
+  }) {
+    let isSanctionedLocation: boolean | undefined;
+    let isSupportedLocation: boolean | undefined;
+    const offering = await this.productConfigurationManager.fetchCMSData(
+      args.offeringId,
+      args.acceptLanguage || undefined,
+      args.selectedLanguage
+    );
+    const supportedCountries = offering.countries.map((country) =>
+      country.slice(0, 2)
+    );
+
+    if (args.countryCode) {
+      isSanctionedLocation = args.sanctionedLocations.includes(
+        args.countryCode
+      );
+      isSupportedLocation = supportedCountries.includes(args.countryCode);
+    }
+
+    if (isSanctionedLocation) {
+      return {
+        status: LocationStatus.SanctionedLocation,
+        message: `Customer is in a sanctioned location: ${args.countryCode}`,
+      };
+    } else if (isSupportedLocation === false) {
+      return {
+        status: LocationStatus.ProductNotAvailable,
+        message: `Product is not available in customer's location: ${args.countryCode}`,
+      };
+    } else {
+      return {
+        status: LocationStatus.Unresolved,
+        message: 'Country code was not resolved',
+      };
```

Comment:
Could this be moved to the top of the method before any of the async logic, i.e. fetchCMSData, gets called?